### ISSUE TITLE
TestNG not published on Maven Central

### ISSUE THREAD

**Issue by skirkpatrick** [cbeust/testng#1013]
Not sure if this is the right place to report this since it's not a code issue, but...

TestNG 6.9.11 was released a few days ago, but it seems it was never published to [Maven central](https://repo.maven.apache.org/maven2/org/testng/testng/). The 6.9.11 package is available on [bintray](https://bintray.com/cbeust/maven/testng/view) and I noticed that the [TestNG documentation](http://testng.org/doc/download.html) demonstrates adding the bintray repo to the POM, but I haven't seen anything about no longer publishing to Maven central so I thought I'd check here before I do that.

**Comment by cbeust** [comment]
Yes, I encountered an error when I tried to deploy to Maven Central (as I often do) so I deferred, since TestNG is already on JCenter.

I've been tempted for a while to make TestNG JCenter-only going forward.

**Comment by juherr** [comment]
> I've been tempted for a while to make TestNG JCenter-only going forward.

Not sure it is a good idea: without maven central, the risk is to have many questions on that problem :(

**Comment by cbeust** [comment]
The problem is just that users who didn't configure JCenter in their build file are stuck on 6.9.10. If they want to upgrade moving forward, they just need to add JCenter to their build file.

Not saying I won't deploy to Maven Central (will try again), just that it doesn't seem to be a huge deal to me.

**Comment by juherr** [comment]
> it doesn't seem to be a huge deal to me.

We are agree but my feeling about users is less optimistic than yours ;)

**Comment by leerho** [comment]
I have added

 <repositories>
    <repository>
      <id>central</id>
      <name>bintray</name>
      <url>http://jcenter.bintray.com</url>
    </repository>
  </repositories>

 to my pom file, but my travis-ci build still fails with

The POM for org.testng:testng:jar:6.9.11 (or 12) is missing, no dependency information available.

Help please.

**Comment by leerho** [comment]
previous message was garbled ... I added the jcenter repository to my pom file: [pom.xml](https://github.com/DataSketches/sketches-core/blob/master/pom.xml) and still get the above error.

**Comment by missedone** [comment]
hi @leerho 
what if you run maven build on your local (with pointing to 6.9.11 or 6.9.12 on jcenter), is everything OK?

**Comment by leerho** [comment]
blockquote, div.yahoo_quoted { margin-left: 0 !important; border-left:1px #715FFA solid !important; padding-left:1ex !important; background-color:white !important; }  Building local seems to work fine. So I was puzzled why the Travis-ci build failed.
Lee


Sent from Yahoo Mail for iPhone


On Tuesday, June 28, 2016, 8:12 PM, Nick Tan <notifications@github.com> wrote:


hi @leerho 
what if you run maven build on your local (with pointing to 6.9.11 or 6.9.12 on jcenter), is everything OK?

**Comment by leerho** [comment]
The travis-ci folks found the error.   I had copy-pasted the recommended addition to the POM from the TestNG Download site: [http://testng.org/doc/download.html](http://testng.org/doc/download.html):

`<repositories>
  <repository>
    <id>central</id>
    <name>bintray</name>
    <url>http://jcenter.bintray.com</url>
  </repository>
</repositories>`

But the ID "central" conflicts with the "central" ID of the master parent POM.  changing it to "jcenter" did the trick.   I would recommend this be changed in the Download document so others don't fall into this trap. 

Lee.

**Comment by cbeust** [comment]
Done, I updated the web site:  http://testng.org/doc/download.html

Thanks!

**Comment by mcekovic** [comment]
Still, do not think it is good idea to go only with JCenter, because many builds use only Maven Central and beside the need to add reference to additional repo, adding additional repo to build slows down dependency checking by factor two (it must go to both repos to check for newer versions).

**Comment by cbeust** [comment]
Smart build systems usually check these repos in parallel and return the first result, though, so I don't think it really slows down your builds.

I encountered some frustrating PGP problems last time I tried to upload TestNG to Maven Central, will try again when I get a minute.

**Comment by leerho** [comment]
It would be greatly appreciated if you could continue to use Maven Central.  It helps simplify the pom if nothing else, and pom files are already huge and overly complex even for simple projects.  :)
 

    On Thursday, July 28, 2016 9:15 AM, Cedric Beust <notifications@github.com> wrote:
 

 Smart build systems usually check these repos in parallel and return the first result, though, so I don't think it really slows down your builds.I encountered some frustrating PGP problems last time I tried to upload TestNG to Maven Central, will try again when I get a minute.—

**Comment by Toilal** [comment]
Many corporate private repository restrict to maven central only. It will be blocking for many users, I think ...

**Comment by leerho** [comment]
I'm not sure if this has been reported but 6.9.12 is missing some dependencies in its POM and Maven is unable to resolve them.  I get this trace:

java.lang.NoClassDefFoundError: com/google/common/primitives/Ints
	at org.testng.internal.annotations.JDK15TagFactory.createDataProviderTag(JDK15TagFactory.java:335)
	at org.testng.internal.annotations.JDK15TagFactory.createTag(JDK15TagFactory.java:59)
	at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:217)
	at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:111)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:326)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:261)
	at org.testng.internal.Parameters.handleParameters(Parameters.java:418)
	at org.testng.internal.Invoker.handleParameters(Invoker.java:1240)
	at org.testng.internal.Invoker.createParameters(Invoker.java:980)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1070)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:746)
	at org.testng.TestRunner.run(TestRunner.java:600)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:366)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:361)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:319)
	at org.testng.SuiteRunner.run(SuiteRunner.java:268)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1264)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1189)
	at org.testng.TestNG.runSuites(TestNG.java:1104)
	at org.testng.TestNG.run(TestNG.java:1076)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:165)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:78)

**Comment by leerho** [comment]
I'm not sure if this has been reported but 6.9.12 is missing some dependencies in its POM and Maven is unable to resolve them.  I get this trace:

java.lang.NoClassDefFoundError: com/google/common/primitives/Ints
	at org.testng.internal.annotations.JDK15TagFactory.createDataProviderTag(JDK15TagFactory.java:335)
	at org.testng.internal.annotations.JDK15TagFactory.createTag(JDK15TagFactory.java:59)
	at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:217)
	at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:111)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:326)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:261)
	at org.testng.internal.Parameters.handleParameters(Parameters.java:418)
	at org.testng.internal.Invoker.handleParameters(Invoker.java:1240)
	at org.testng.internal.Invoker.createParameters(Invoker.java:980)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1070)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:746)
	at org.testng.TestRunner.run(TestRunner.java:600)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:366)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:361)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:319)
	at org.testng.SuiteRunner.run(SuiteRunner.java:268)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1264)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1189)
	at org.testng.TestNG.runSuites(TestNG.java:1104)
	at org.testng.TestNG.run(TestNG.java:1076)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:165)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:78)

**Comment by leerho** [comment]
I'm not sure if this has been reported but 6.9.12 is missing some dependencies in its POM and Maven is unable to resolve them.  I get this trace:

java.lang.NoClassDefFoundError: com/google/common/primitives/Ints
	at org.testng.internal.annotations.JDK15TagFactory.createDataProviderTag(JDK15TagFactory.java:335)
	at org.testng.internal.annotations.JDK15TagFactory.createTag(JDK15TagFactory.java:59)
	at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:217)
	at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:111)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:326)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:261)
	at org.testng.internal.Parameters.handleParameters(Parameters.java:418)
	at org.testng.internal.Invoker.handleParameters(Invoker.java:1240)
	at org.testng.internal.Invoker.createParameters(Invoker.java:980)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1070)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:746)
	at org.testng.TestRunner.run(TestRunner.java:600)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:366)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:361)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:319)
	at org.testng.SuiteRunner.run(SuiteRunner.java:268)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1264)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1189)
	at org.testng.TestNG.runSuites(TestNG.java:1104)
	at org.testng.TestNG.run(TestNG.java:1076)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:165)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:78)

**Comment by juherr** [comment]
I reopen the issue because I think it is still relevant.

**Comment by rliesenfeld** [comment]
I just now discovered that TestNG artifacts are not being deployed to Maven Central anymore, so here are my two cents...

My OSS project (jmockit.org) has a dependency on TestNG, and I don't feel good about adding the "jcenter" repository to my pom.xml, just for TestNG. I don't believe that JCenter will ever really compete with Maven Central.

Cedric says it often had problems to deploy TestNG there, Maybe there is something that can be improved in TestNG's pom.xml? The JMockit library is regularly deployed to Central, and I haven't had any problems. So, feel free to examine it for ideas (it uses "maven-gpg-plugin" and "nexus-staging-maven-plugin", nothing more - I then simply run "maven deploy").

**Comment by juherr** [comment]
@rliesenfeld Thanks for your proposition. I'm sure @cbeust will appreciate the help.
But just notice that we dropped maven (not yet removed) as build tool and is currently replaced by gradle (maybe by kobalt later).

**Comment by dpinol** [comment]
I agree that it should be publihed at maven central because it's the defacto standard. For instance, I had not realized there was a newer version because I just saw 6.9.10 there

**Comment by cbeust** [comment]
I just deployed 6.9.13.3 to Sonatype, it should be available on Maven Central in the next few hours, I just had to add a few tweaks to the [`Build.kt`](https://github.com/cbeust/testng/commit/95d748af4385aba73af7be7c1e1e972842fa3980) and after that, JCenter successfully synced:

![image](https://cloud.githubusercontent.com/assets/92322/18756929/95811d9e-80a6-11e6-8ab5-ec00dd454c4f.png)

**Comment by juherr** [comment]
Awesome!! 👍 Just have to manage optional and provided dependencies, and everything will be perfect! :D

**Comment by cbeust** [comment]
@juherr Can you elaborate about `optional` and `provided`?

**Comment by juherr** [comment]
Some dependencies are declared optional or provided. The generated pom should contain this information. 
I had to use a specific plugin to be able to declare the information and have the expected pom. 
Maybe an extra plugin should be done for kobalt too.

**Comment by cbeust** [comment]
Should be fixed. Can you take a look on JCenter before I sync to Maven?

https://bintray.com/cbeust/maven/testng/6.9.13.4#files/org/testng/testng/6.9.13.4

-- 
Cédric

**Comment by juherr** [comment]
Great, everything looks good now! Even the central sync! :D

**Comment by cbeust** [comment]
Ok I just synced 6.9.13.4.

-- 
Cédric

**Comment by juherr** [comment]
:+1: Don't forget tags (is it possible to do it automatically with kobalt?)
